File: src/android/com/phemium/sipvideocall/CallManager.java
package com.phemium.sipvideocall;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Rect;
import android.util.Log;

import com.phemium.sipvideocall.constant.Constant;
import com.phemium.sipvideocall.data.CallData;
import com.phemium.sipvideocall.data.LanguageResource;
import com.phemium.sipvideocall.listener.BandwidthListener;
import com.phemium.sipvideocall.listener.CallManagerListener;
import com.phemium.sipvideocall.listener.LinphoneManagerListener;
import org.apache.cordova.CordovaPlugin;

import java.util.Timer;
import java.util.TimerTask;

import static com.phemium.sipvideocall.constant.Constant.SWIPE_BOTTOM;
import static com.phemium.sipvideocall.constant.Constant.SWIPE_RIGHT;

/**
 * Created by Tom on 1/5/2017.
 */

public class CallManager extends Object implements LinphoneManagerListener, BandwidthListener {

    public CallData mCallData;
    public CallManagerListener mCallManagerListener;
    public CordovaPlugin mCordovaPlugin;

    private Context cordovaContext;
    private LinphoneManager mLinphoneManager;
    private Boolean bCameraPermissionAllowed = true;
    private final static String TAG = "CallManager";
    private static boolean isCalling = false;
    private boolean isCallConnected = false;
    public boolean bIncoming = false;
    public String logPath = "";
    private boolean bRetry = false;
    private boolean checkNetwork = false;
    private boolean registable = true;

    public void startCall(CallData callData, CallManagerListener listener, CordovaPlugin cordovaPlugin) {
        if (CallManager.isCalling){
            Log.e(TAG, "Call is already started.");
            return;
        }
        Log.e(TAG, "Start Call");
        this.mCallData = callData;
        this.mCallManagerListener = listener;
        this.mCordovaPlugin = cordovaPlugin;
        this.cordovaContext = cordovaPlugin.cordova.getActivity();
        LanguageResource.getInstance().readFromJsonFile(mCallData.language, cordovaContext);
        bRetry = false;
        checkNetwork = false;
        checkPermissions();
//        initializeCall();
    }

    public void checkNetworkStatus(CallData callData, CallManagerListener listener, CordovaPlugin cordovaPlugin){
        this.mCallData = callData;
        this.mCallManagerListener = listener;
        this.mCordovaPlugin = cordovaPlugin;
        this.cordovaContext = cordovaPlugin.cordova.getActivity();
        LanguageResource.getInstance().readFromJsonFile(mCallData.language, cordovaContext);

        checkNetwork = true;
        Log.e(TAG, "Check Network");
        initializeCall();
    }

    public void reOpen() {
        mLinphoneManager.reOpen();
    }

    public void hangUp() {
        mLinphoneManager.hangup();
    }

    public void onChatMessageArrived() {
        mLinphoneManager.onChatMessageArrived();
    }

    public void setOverlayBoundary(Rect newRect) {
        mLinphoneManager.setOverlayBoundary(newRect);
    }

    public void changeOrientation(){
        mLinphoneManager.changeOrientation();
    }

    private void checkPermissions() {
        Log.e(TAG, "checkPermissions called.");
        if (mCordovaPlugin != null) {
            if (!mCordovaPlugin.cordova.hasPermission(Manifest.permission.RECORD_AUDIO)) {
                Log.e(TAG, "Hasn't Record Audio Permission.");
                if (!mCordovaPlugin.cordova.hasPermission(Manifest.permission.CAMERA)) {
                    mCordovaPlugin.cordova.requestPermissions(mCordovaPlugin, Constant.PERMISSION_CHECK_MEDIA_BEFORE_CALL, new String[]{Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO});
                } else {
                    mCordovaPlugin.cordova.requestPermission(mCordovaPlugin, Constant.PERMISSION_CHECK_MEDIA_BEFORE_CALL, Manifest.permission.RECORD_AUDIO);
                }
            } else {
                Log.e(TAG, "Has Record Audio Permission.");
                if (!mCordovaPlugin.cordova.hasPermission(Manifest.permission.CAMERA)) {
                    bCameraPermissionAllowed = false;
                } else {
                    bCameraPermissionAllowed = true;
                }
                initializeCall();
            }
        } else {
            initializeCall();
        }
    }

    public void onRequestPermissionResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == Constant.PERMISSION_CHECK_MEDIA_BEFORE_CALL) {
            if (grantResults.length <= 0) {
                onCallFailed(Constant.CALL_RESULT_ERROR_RECORD_PERMISSION_NOT_ALLOWED , "Record Permission Not Allowed"); // requesting permission is interrupted by the user
                return;
            }
            int i = Utils.indexOfStringArray(permissions, Manifest.permission.RECORD_AUDIO);
            if (i > 0) {
                if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    int j = Utils.indexOfStringArray(permissions, Manifest.permission.CAMERA);
                    if (j > 0) {
                        if (grantResults[j] == PackageManager.PERMISSION_GRANTED) {
                            bCameraPermissionAllowed = true;
                        } else {
                            bCameraPermissionAllowed = false;
                        }
                    } else {
                        bCameraPermissionAllowed = true;
                    }
                    initializeCall();

                } else {
                    onCallFailed(Constant.CALL_RESULT_ERROR_RECORD_PERMISSION_NOT_ALLOWED, "Record Permission Not Allowed"); // microphone permission not allowed.
                }
            }
        }
    }

    private boolean checkAndroidVersion() {
        return (android.os.Build.VERSION.SDK_INT > android.os.Build.VERSION_CODES.LOLLIPOP_MR1);
    }

    public void initializeCall() {
        isCallConnected = false;
        initializeLinphone();
        configureSettings();
        mLinphoneManager.bIncomingCall = bIncoming;
        mLinphoneManager.badgeNum = 0;
        bIncoming = true;   // When Call retry, shows connecting screen.

        mLinphoneManager.setCheckNetwork(checkNetwork);
        mLinphoneManager.doRegister();
    }

    private void initializeLinphone() {
        LinphoneManager linphoneManager = LinphoneManager.getInstance();
        mLinphoneManager = linphoneManager;
        mLinphoneManager.callData = mCallData;
        if (!bRetry) {
            mLinphoneManager.logsStatus = "";
            mLinphoneManager.connecting_failed_cnt = 0;
        }
        mLinphoneManager.startLibLinphone(cordovaContext, this);
    }

    private void configureSettings() {
        mLinphoneManager.bCameraPermissionAllowed = bCameraPermissionAllowed;
        mLinphoneManager.clearAuthInfoAndProxyConfig();
        mLinphoneManager.enableLogCollection(mCallData.logEnable);
        mLinphoneManager.setProxy();
        mLinphoneManager.setTurnSetting();
        mLinphoneManager.setUser();
        mLinphoneManager.setCallQualityParams();
        mLinphoneManager.setTransportValue();
        mLinphoneManager.setVideoSize();
    }

    @Override
    public void onRegisterSucceeded() {
        Log.e(TAG, "Register Success");
        if (checkNetwork){
            registable = true;
//            mLinphoneManager.doUnRegister();
            NetworkManager.getBandwidth(cordovaContext, this);
            return;
        }
        if (!isCallConnected){
            isCallConnected = true;
            Log.e(TAG, "Call Connected");

            Timer callStartTimer = new Timer();
            callStartTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    if (!mCallData.onlyAudio) {
                        mLinphoneManager.call(mCallData.toAddress);
                    } else {
                        mLinphoneManager.audioCall(mCallData.toAddress);
                    }
                }
            }, 2500);
        }
    }

    @Override
    public void onRegisterFailed(String workflow) {
        Log.e(TAG, "Register Failed");
        if (checkNetwork){
            registable = false;
            NetworkManager.getBandwidth(cordovaContext, this);
            return;
        }
    }

    @Override
    public void onCallFailed(int code , String workflow) {

        if (mCallManagerListener != null) {
//            mLinphoneManager.destroyLibLinphone();
            Utils.insertCallLog(cordovaContext, mCallData.consultantName, mLinphoneManager.bIncomingCall, mLinphoneManager.nCallDuration);
            mCallManagerListener.onThrowError(code, workflow, mLinphoneManager.logPath);

            mLinphoneManager.sendLinphoneDebug();
        }
    }

    @Override
    public void onCallReleased(String workflow) {
        mLinphoneManager.badgeNum = 0;
        mLinphoneManager.swipeHState = SWIPE_BOTTOM;
        mLinphoneManager.swipeWState = SWIPE_RIGHT;

        if (mCallManagerListener != null) {

//            mLinphoneManager.destroyLibLinphone();
            Utils.insertCallLog(cordovaContext, mCallData.consultantName,  mLinphoneManager.bIncomingCall, mLinphoneManager.nCallDuration);
            mCallManagerListener.onCallReleased(workflow , mLinphoneManager.logPath);

            mLinphoneManager.sendLinphoneDebug();
        }
    }

    @Override
    public void onCallEventOccur(int eventCode) {
        if (mCallManagerListener != null) {
            mCallManagerListener.onEventComeUp(eventCode);
        }
    }

    @Override
    public void onCallMinimized(int callDuration) {
        if (mCallManagerListener != null) {
            mCallManagerListener.onEventMinimized(callDuration);
        }
    }

    @Override
    public void onCallSendQuality(String quality) {
        if (mCallManagerListener != null) {
            mCallManagerListener.onEventSendQuality(quality);
        }
    }

    @Override
    public void onCallRetry() {
        bRetry = true;
        initializeCall();
    }

    @Override
    public void onCallRinging() {

    }

    @Override
    public void onBandwidthCalcuated(float bandwidth) {
        mCallManagerListener.onCheckReleased(registable, NetworkManager.getNetworkType(cordovaContext), bandwidth);
    }
}

File: src/ios/CallManager.m
//
//  CallManager.m
//  SipVideoPluginTest
//
//  Created by SCNDev1 on 12/22/16.
//
//

#import "CallManager.h"
#import "LinPhoneManager.h"
#import "CallData.h"
#import "AppDelegate+callprovider.h"
#import "PublicData.h"
#import "StringUtil.h"
#import "Constant.h"
#import "ProviderDelegate.h"

#import <AVFoundation/AVFoundation.h>

@interface CallManager() <LinphoneManagerListener>{
    double startRegisterTime;
    BOOL isRegistered;
    BOOL isCallConnected;
    BOOL isIncoming;
    BOOL isCallKitNeeded;
    NSUUID *incomingCallUUID;
    BOOL isRetry;
    BOOL isAllowCameraPermission;
    BOOL checkNetwork;
}

@end

@implementation CallManager

@synthesize callData;
@synthesize delegate;
@synthesize linphoneManager;

- (instancetype)init {
    if (self = [super init]) {
        isIncoming = NO;
        isCallKitNeeded = YES;
        isAllowCameraPermission = NO;
        [PublicData sharedInstance].callManager = nil;
    }
    return self;
}

- (void)startCall:(CallData *)callDt callManagerDelegate:(id<CallManagerDelegate>)callManagerDelegate {
    
    if ([PublicData sharedInstance].isCalling == YES) {
        NSLog(@"CallManager Call is already stared.");
        return;
    }
    NSLog(@"CallManager Start Call");
    
    callData = callDt;
    delegate = callManagerDelegate;
    [StringUtil setLanguage:callData.language];
    isRetry = false;
    checkNetwork = false;
    [self checkPermissions];
//    [self initializeCall];
}

-(void)checkNetworkStatus: (CallData *)callDt callManagerDelegate:(id<CallManagerDelegate>)callManagerDelegate {
    
    return;
    callData = callDt;
    delegate = callManagerDelegate;
    [StringUtil setLanguage:callData.language];
    
    checkNetwork = true;
    [self initializeCall];
}

- (void)reOpen {
    if (self.linphoneManager) {
        [self.linphoneManager reOpen];
    }
}

- (void)hangUp {
    if (self.linphoneManager) {
        [self.linphoneManager hangupWithCause:CALL_HANGUP_BY_CALLER];
    }
}

- (void)onChatMessageArrived {
    if (self.linphoneManager) {
        [self.linphoneManager onChatMessageArrived];
    }
}

- (void)startSimulatingIncomingCallWithCallKit:(CallData *)callDt callManagerDelegate:(id<CallManagerDelegate>)callManagerDelegate {
    [PublicData sharedInstance].incomingCallData = callDt;
    UIBackgroundTaskIdentifier backgroundTaskIdentifier = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        [((AppDelegate*)[UIApplication sharedApplication].delegate) displayIncomingCall:[NSUUID UUID] handle:@"Unknown name" hasVideo:(callDt.onlyAudioCall==1 ? NO : YES) completion:^{
            [[UIApplication sharedApplication] endBackgroundTask:backgroundTaskIdentifier];
        }];
    });
}

- (void)onAnswerIncomingCall:(NSUUID *)callUuid {
    if ([PublicData sharedInstance].incomingCallData) {
        isIncoming = YES;
        incomingCallUUID = callUuid;
        [self startCall:[PublicData sharedInstance].incomingCallData callManagerDelegate:(AppDelegate*)([UIApplication sharedApplication].delegate)];
    }
}

- (void)onAnswerIncomingCallWithoutCallKit:(CallData *)callDt {
    isIncoming = YES;
    isCallKitNeeded = NO;
    [self startCall:callDt callManagerDelegate:(AppDelegate*)([UIApplication sharedApplication].delegate)];
}

- (void) checkPermissions {
    [self checkAudioPermission];
}

- (void)setIsIncoming :(BOOL) incoming{
    isIncoming = incoming;
}

- (void) checkAudioPermission {
    AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeAudio];
    if(authStatus == AVAuthorizationStatusAuthorized) {
        [self checkVideoPermission];
    } else if(authStatus == AVAuthorizationStatusDenied){
        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    } else if(authStatus == AVAuthorizationStatusRestricted){
        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    } else if(authStatus == AVAuthorizationStatusNotDetermined){
        [self checkVideoPermission];
    } else {
        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    }
}

- (void) checkVideoPermission {
    isAllowCameraPermission = NO;
    //Not check video permission. We can only show doctor's screen if camera permission is denied.
    AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if(authStatus == AVAuthorizationStatusAuthorized) {
//        [self initializeCall];
        isAllowCameraPermission = YES;
    } else if(authStatus == AVAuthorizationStatusDenied){
//        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    } else if(authStatus == AVAuthorizationStatusRestricted){
//        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    } else if(authStatus == AVAuthorizationStatusNotDetermined){
//        [self initializeCall];
        
    } else {
//        [self throwError:CALL_ERROR_PERMISSIONS_DENIED];
    }
    
    [self initializeCall];
}


- (void) initializeCall {
    //Initalize parameters;
    
    [PublicData sharedInstance].isCalling = YES;
    
    startRegisterTime = [[NSDate date] timeIntervalSince1970];
    isRegistered = NO;
    isCallConnected = NO;

    [self initializeLinphone];

    //Set Current Parameters
    [self setLinphoneCallData];
    self.linphoneManager.isCallKitNeeded = isCallKitNeeded;
    self.linphoneManager.isCameraEnable = isAllowCameraPermission;
    self.linphoneManager.currentCallUuid =  [NSUUID UUID];
    if (isIncoming) {
        self.linphoneManager.isIncoming = YES;
        
    } else {
        self.linphoneManager.isIncoming = NO;
    }
    isIncoming = YES;   //when call retry, shows connecting screen
    
    [linphoneManager setCheckNetwork:checkNetwork];
    [linphoneManager doRegister];
}

- (void)initializeLinphone {
    NSLog(@"SipVideoCall:initialize");
        LinphoneManager *manager = [LinphoneManager instance];

    self.linphoneManager = manager;
    self.linphoneManager.listener = self;
    self.linphoneManager.currentCallData = callData;
    self.linphoneManager.badgeNum = 0;
    if (!isRetry) {
        self.linphoneManager.logsStatus = @"";
    }
    [self.linphoneManager startLibLinphone];
}

- (void) setLinphoneCallData {
    // Set Credentials
    NSLog(@"SipVideoCall:setCredentials");
    assert(self);
    [linphoneManager setProxy: callData.proxy address:callData.address];
    [linphoneManager setUser: callData.username pass: callData.password domain:callData.domain forAddress: callData.address];
    [linphoneManager setTurnServer:callData.turnServer0 
        domain0:callData.turnDomain0 username0:callData.turnUsername0 password0:callData.turnPassword0 
        alternate:callData.turnServer1
        domain1:callData.turnDomain1 username1:callData.turnUsername1 password1:callData.turnPassword1
        ];
    [linphoneManager setCallQualityParams:callData.download_bandwidth upload_bandwidth:callData.upload_bandwidth framerate:callData.framerate];
    
    [linphoneManager setTransportValue];
    [linphoneManager setVideoSize];
    [linphoneManager enableLogCollection:callData.log_enable];
}

- (void) onAudioSessionActivated {
    NSLog(@"onAudioSessionActivated called.");
    if (isIncoming) {
        if (callData.onlyAudioCall == 1) {
            [self.linphoneManager audioCall:callData.to];
        } else {
            [self.linphoneManager call:callData.to];
        }
    }
}

- (void)setCredentials {

}

- (void)throwError: (int)error {
    if(self.linphoneManager)
        [self.linphoneManager destroyLibLinphone];
    [delegate onThrowError: error workflow:@"Could not place call because michrophone permissions are denied" logPath:@""];
}

- (void)onRegisterSucceeded {
    NSLog(@"onRegisterSucceed called.");
    if (checkNetwork) {
        [delegate onCheckReleased:true networkType:@"" bandwidth:0];
        return;
    }
    if (isCallConnected == NO) {
        isCallConnected = YES;
        NSLog(@"call connected");
        // else {
        if (callData.onlyAudioCall == 1) {
            [self.linphoneManager audioCall:callData.to];
        } else {
            [self.linphoneManager call:callData.to];
        }
        // }
    }
}

- (void)onRegisterFailedWithMessage: (NSString *) message {
    NSLog(@"Register failed.");
    if(self.linphoneManager)
        [self.linphoneManager destroyLibLinphone];
    [delegate onThrowError: CALL_ERROR_REGISTER_FAILURE workflow:message logPath:self.linphoneManager.logPath];
}

//called when outgoing call is connected
- (void)onCallConnectedWithPeer: (NSUUID *)uuid hasVideo:(Boolean) hasVideo{
    isCallConnected = YES;
    
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    
    [appDelegate informCallProviderCallConnected:uuid hasVideo:hasVideo];
}

- (void)onCallRejectedWithCode: (NSNumber *) code {
}

- (void)onCallFailed: (int)code workflow:(NSString *)workflow{
    [PublicData sharedInstance].isCalling = NO;
//    if(self.linphoneManager)
//        [self.linphoneManager destroyLibLinphone];
    if (self.linphoneManager) {
        [self.linphoneManager sendLinphoneDebug];
    }
    [delegate onThrowError: code workflow:workflow logPath:self.linphoneManager.logPath];
}

- (void)onCallReleased: (NSString * )workflow {
    [PublicData sharedInstance].isCalling = NO;
    linphoneManager.badgeNum = 0;
//    if(self.linphoneManager)
//        [self.linphoneManager destroyLibLinphone];
    if (self.linphoneManager) {
        [self.linphoneManager sendLinphoneDebug];
    }
    [delegate onCallReleased: workflow logPath:self.linphoneManager.logPath];
}

- (void)onCallRinging {

}

- (void)onCallEventOccur: (int)code {
    [delegate onEventComesUp: code];
}

- (void)onCallMinimized:(long)duration{
    [delegate onMinimized: duration];
}

-(void)onCallSendQuality:(NSString *)quality{
    [delegate onSendQuality:quality];
}

- (void)onCallRetry{
    isRetry = true;
    [self initializeCall];
}

- (void)onIncomingCallFrom: (NSString *) from {

}

- (void)onTransferRequestedTo: (NSString *) to {

}

@end
